libraries/shared/storage/backfiller.go

```diff
@@ -0,0 +1,165 @@
+		}
+		blockRange := make([]uint64, 0, nextBinStart-bf.startingBlock+1)
+		for j := bf.startingBlock; j < nextBinStart; j++ {
+			blockRange = append(blockRange, j)
+		}
+		bf.startingBlock = nextBinStart
+		blockRangeBins[i] = blockRange
+	}
+
+	// int64 for atomic incrementing and decrementing to track the number of active processing goroutines we have
+	var activeCount int64
+	// channel for processing goroutines to signal when they are done
+	processingDone := make(chan [2]uint64)
+	forwardDone := make(chan bool)
+
+	// for each block range bin spin up a goroutine to batch fetch and process state diffs for that range
+	go func() {
+		for _, blockHeights := range blockRangeBins {
+			// if we have reached our limit of active goroutines
+			// wait for one to finish before starting the next
+			if atomic.AddInt64(&activeCount, 1) > defaultMaxBatchNumber {
+				// this blocks until a process signals it has finished
+				<-forwardDone
+			}
+			go func(blockHeights []uint64) {
+				payloads, fetchErr := bf.fetcher.FetchStateDiffsAt(blockHeights)
+				if fetchErr != nil {
+					errChan <- fetchErr
+				}
+				for _, payload := range payloads {
+					stateDiff := new(statediff.StateDiff)
+					stateDiffDecodeErr := rlp.DecodeBytes(payload.StateDiffRlp, stateDiff)
+					if stateDiffDecodeErr != nil {
+						errChan <- stateDiffDecodeErr
+						continue
+					}
+					accounts := utils.GetAccountsFromDiff(*stateDiff)
+					for _, account := range accounts {
+						logrus.Trace(fmt.Sprintf("iterating through %d Storage values on account", len(account.Storage)))
+						for _, storage := range account.Storage {
+							diff, formatErr := utils.FromGethStateDiff(account, stateDiff, storage)
+							if formatErr != nil {
+								errChan <- formatErr
+								continue
+							}
+							logrus.Trace("adding storage diff to results",
+								"keccak of address: ", diff.HashedAddress.Hex(),
+								"block height: ", diff.BlockHeight,
+								"storage key: ", diff.StorageKey.Hex(),
+								"storage value: ", diff.StorageValue.Hex())
+							backFill <- diff
+						}
+					}
+				}
+				// when this goroutine is done, send out a signal
+				processingDone <- [2]uint64{blockHeights[0], blockHeights[len(blockHeights)-1]}
+			}(blockHeights)
+		}
+	}()
+
+	// goroutine that listens on the processingDone chan
+	// keeps track of the number of processing goroutines that have finished
+	// when they have all finished, sends the final signal out
+	go func() {
+		goroutinesFinished := 0
+		for {
+			select {
+			case doneWithHeights := <-processingDone:
+				atomic.AddInt64(&activeCount, -1)
+				select {
+				// if we are waiting for a process to finish, signal that one has
+				case forwardDone <- true:
+				default:
+				}
+				logrus.Infof("finished fetching gap sub-bin from %d to %d", doneWithHeights[0], doneWithHeights[1])
+				goroutinesFinished++
+				if goroutinesFinished == int(numberOfBins) {
+					done <- true
+					return
+				}
+			}
+		}
+	}()
```

Comment:
Any way to extract some functions to abstract away implementation details of what's happening here? A little tough to grok the whole block

Reply:
Yes this is a bit unwieldy :ok_hand: